Secure Local Development

davidpatrick opened the pull request:
Setting up a secure local development server

davidpatrick committed:
Secure Local Development

joshcanhelp commented on articles/libraries/secure-local-development.md:
```diff
@@ -0,0 +1,68 @@
+---
+
+# Secure Local Development
+Google is introducing new restrictions on cookie usage in version 80 of Chrome with a new cookie attribute named [Same Site](https://auth0.com/blog/browser-behavior-changes-what-developers-need-to-know/).  This article covers how it impacts using Auth0 in your local environment with certain SDKs.
```
I would leave out anything related to specific browsers as they will all end up implementing the same changes. Also, SameSite itself is not new, only the changes to how browsers are handling it.

joshcanhelp commented on articles/libraries/secure-local-development.md:
```diff
@@ -0,0 +1,68 @@
+# Secure Local Development
+Google is introducing new restrictions on cookie usage in version 80 of Chrome with a new cookie attribute named [Same Site](https://auth0.com/blog/browser-behavior-changes-what-developers-need-to-know/).  This article covers how it impacts using Auth0 in your local environment with certain SDKs.
+
+This article only impacts you if you are using one of the following SDKS:
```
This part we definitely want to leave out. In the end, using HTTPS on your local environment does other things besides make SameSite=none function:

- It keeps communication between Auth0 and the localhost server secret as well. If you're testing with production or sensitive data then encrypted transfer is important no matter what the receiving server is. 
- It tests a very critical part of the development stack; if you test without secure channels on localhost but run on secure channels in production, you're missing a potentially large source of issues. 
- It enforces the fact that these protocols should happen over secure channels.

joshcanhelp commented on articles/libraries/secure-local-development.md:
```diff
@@ -0,0 +1,68 @@
+- express-openid-connect
+- another one
+
+These SDKs utilize cookies to handle request state and prevent replay attacks.  When the authentication request is handed back from Auth0 to your local server, the new same site restrictions will prevent the cookie being passed if the local server is not secure (https).  Without this cookie value, the local server will be unable to complete the authentication request.
```
```suggestion
Our web application SDKs use cookies to handle the transient data needed for authentication. When the authentication response is returned from Auth0 to your application, certain required cookie attributes, like `SameSite` and `Secure`, might prevent this data from being passed if the application is not served over secure channels (URLs using `https`). Without this data, the application will refuse the authentication request and the login will fail. 
```

joshcanhelp replied:
I would put a note about the browser changes here but leave out specific ones. Might be good to link to our blog post about it as well in case people want to learn more:

https://auth0.com/blog/browser-behavior-changes-what-developers-need-to-know/#When-

joshcanhelp commented on articles/libraries/secure-local-development.md:
```diff
@@ -0,0 +1,68 @@
+These SDKs utilize cookies to handle request state and prevent replay attacks.  When the authentication request is handed back from Auth0 to your local server, the new same site restrictions will prevent the cookie being passed if the local server is not secure (https).  Without this cookie value, the local server will be unable to complete the authentication request.
+
+## How to set up a secure local environment
+Instructions on how to set up a secure local environment can be daunting and burdensome.  However, thanks to the open source community, [Mkcert](https://github.com/FiloSottile/mkcert) is a tool that automates and simplifies the implementation.
```
Second sentence is a bit awkward ...

joshcanhelp replied:
Maybe give a quick introduction to the process and then numbered headers for the main segments? The process is generally:

1. Generating a certificate on your machine that can be trusted (root cert)
2. Generating a certificate for your application domain based on that certificate
3. Loading that certificate when the application is loaded

I found that understanding the very basic steps of the process helped with troubleshooting and expanded my knowledge on the whole process. As long as we can describe it in simple, accurate terms, I think developers will be better off overall.

joshcanhelp commented on articles/libraries/secure-local-development.md:
```diff
@@ -0,0 +1,68 @@
+Follow the [installation instructions](https://github.com/FiloSottile/mkcert#installation) for your specific operating system.
+
+### Installing the local Certificate Authority
+After you have mkcert installed you can use it to install the local Certificate Authority.  The local Certificate Authority will allow for mkcert to generate locally-trusted certificates and accompanying keys to allow your local development server to run on `https`.
```
I think if we use "Certificate Authority," we should explain what we're talking about. We kind of just dropped the term here without explanation. Also, we're talking about installing twice, which we can probably explain a bit better.

joshcanhelp commented on articles/libraries/secure-local-development.md:
````diff
@@ -0,0 +1,68 @@
+```
+
+### Generate a locally-trusted certificate & key
+The next step is to generate the locally-trusted certificate and key for use on your local development server. This example will assume you are running your local server on `https://localhost:{port}` or `https://127.0.0.1:{port}`.  Multiple values can be passed into mkcert to support a range of address names.
````
I wonder if if would be worth it to add a note/panel to explain a bit more about what's going on here with the domains. Maybe also make note of the fact that the localhost one will cover quickstarts.

joshcanhelp commented on articles/libraries/secure-local-development.md:
```diff
@@ -0,0 +1,68 @@
+const https = require('https');
+const app = express();
+
+https.createServer({key: key, cert: cert }, app).listen(port, () => {
```
`port` is undefined here

joshcanhelp reviewed (changes requested):
Great start here! Seeing it on the docs page, it's pretty darn straight-forward. I'm leaning towards a bit more explanation about what's happening, if only because this is documentation and not a quickstart.

davidpatrick committed:
Clarifying instructions

joshcanhelp commented on articles/libraries/secure-local-development.md:
```diff
@@ -0,0 +1,74 @@
+---
+
+# Secure Local Development
+Commonly developers run local web servers on non-secure channels (http).  Generally because the perceived difficulty of running a secure local server (https), and/or it is not a requirement for their local setup.  This guide will discuss when you should run a secure local server `https` and provide simple instructions for setting it up.
```
First two sentences here need work, seems like it should be a single sentence.

joshcanhelp replied:
Also, you probably want to settle on backticks or no for http/`http`/https/`https`

joshcanhelp commented on articles/libraries/secure-local-development.md:
```diff
@@ -0,0 +1,74 @@
+Commonly developers run local web servers on non-secure channels (http).  Generally because the perceived difficulty of running a secure local server (https), and/or it is not a requirement for their local setup.  This guide will discuss when you should run a secure local server `https` and provide simple instructions for setting it up.
+
+## When to use a secure local server
+Testing locally on non-secure channels `http` is generally safe for servers that don't communicate with external services.  However, when your local server is communicating with external services (ie. Auth0), we recommend running your local server on `https` for the following reasons:
```
👍

joshcanhelp replied:
👇 Bullet points as well 👍👍👍

joshcanhelp commented on articles/libraries/secure-local-development.md:
```diff
@@ -0,0 +1,74 @@
+- Cookies using secure or [same-site](https://auth0.com/blog/browser-behavior-changes-what-developers-need-to-know/) will not be sent to local `http` disrupting authentication and other functionality.
+
+## How to set up a secure local server
+When your web browser visits a secure web page (https) the browser verifies the SSL certificate supplied by the server with a Certificate Authority.
```
```suggestion
When you visit a secure web page (served over https), your browser will verify the SSL certificate supplied by the server with a Certificate Authority. When loading a secure web page from a local server, you can create an authority just for your machine and generate certificates that only your browser will trust.
```

joshcanhelp commented on articles/libraries/secure-local-development.md:
```diff
@@ -0,0 +1,74 @@
+## How to set up a secure local server
+When your web browser visits a secure web page (https) the browser verifies the SSL certificate supplied by the server with a Certificate Authority.
+
+This can be handled locally in 3 easy steps:
```
```suggestion
The process to do this is:
```

joshcanhelp commented on articles/libraries/secure-local-development.md:
```diff
@@ -0,0 +1,74 @@
+
+This can be handled locally in 3 easy steps:
+
+1. Install a local Certificate Authority
```
```suggestion
1. Create and install a local Certificate Authority
```

joshcanhelp commented on articles/libraries/secure-local-development.md:
```diff
@@ -0,0 +1,74 @@
+This can be handled locally in 3 easy steps:
+
+1. Install a local Certificate Authority
+2. Generate SSL certificate for the address/domain (ie. localhost)
```
```suggestion
2. Generate an SSL certificate for the domain being used (ie. localhost) using this new Authority.
```

joshcanhelp commented on articles/libraries/secure-local-development.md:
```diff
@@ -0,0 +1,74 @@
+2. Generate SSL certificate for the address/domain (ie. localhost)
+3. Serve the SSL certificate from your web application
+
+To get started, download [Mkcert](https://github.com/FiloSottile/mkcert) and follow the [installation instructions](https://github.com/FiloSottile/mkcert#installation) for your specific operating system.
```
Will this get lost if someone is scanning?

joshcanhelp reviewed (changes requested):
Much better! I have a few suggestions here, then let's post to the internal channel to see if anyone else has feedback. I'm close to releasing the Express changes for SameSite and would like to include this guide in the release notes.

davidpatrick committed:
Update articles/libraries/secure-local-development.md

Co-Authored-By: Josh Cunningham <josh.cunningham@auth0.com>

davidpatrick committed:
Update articles/libraries/secure-local-development.md

Co-Authored-By: Josh Cunningham <josh.cunningham@auth0.com>

davidpatrick committed:
Update articles/libraries/secure-local-development.md

Co-Authored-By: Josh Cunningham <josh.cunningham@auth0.com>

davidpatrick committed:
Update articles/libraries/secure-local-development.md

Co-Authored-By: Josh Cunningham <josh.cunningham@auth0.com>

davidpatrick committed:
Update

- added mkcert install as a listed step
- rephrased top paragrah
- shortened one of the bulletpoints in when to use a secure local server

joshcanhelp commented on articles/libraries/secure-local-development.md:
```diff
@@ -0,0 +1,75 @@
+---
+
+# Secure Local Development
+Local devleopment environments generally run on non-secure channels (ie: `http://localhost`) out of the box.  This guide will discuss when you should run a secure local server and how to setup `https` on localhost.
```
```suggestion
Local development environments typically run on non-secure channels (ie: `http://localhost`) out of the box.  This guide will discuss when you should run a secure local server and how to setup `https` on localhost.
```

joshcanhelp commented on articles/libraries/secure-local-development.md:
```diff
@@ -0,0 +1,75 @@
+Local devleopment environments generally run on non-secure channels (ie: `http://localhost`) out of the box.  This guide will discuss when you should run a secure local server and how to setup `https` on localhost.
+
+## When to use a secure local server
+Testing locally on non-secure channels `http` is generally safe for servers that don't communicate with external services.  However, when your local server is communicating with external services (ie. Auth0), we recommend running your local server on `https` for the following reasons:
```
```suggestion
Testing over non-secure channels `http` is generally safe for local servers that don't communicate with external services, like Auth0.  However, when your local server is communicating with external services, we recommend running your local server on `https` for the following reasons:
```

joshcanhelp commented on articles/libraries/secure-local-development.md:
```diff
@@ -0,0 +1,75 @@
+## When to use a secure local server
+Testing locally on non-secure channels `http` is generally safe for servers that don't communicate with external services.  However, when your local server is communicating with external services (ie. Auth0), we recommend running your local server on `https` for the following reasons:
+
+- Communication between an external service (ie. Auth0) and your localhost should be encrypted to protect any sensitive information.
```
```suggestion
- Communication between an external service and your localhost should be encrypted to protect any sensitive information.
```

joshcanhelp commented on articles/libraries/secure-local-development.md:
```diff
@@ -0,0 +1,75 @@
+
+- Communication between an external service (ie. Auth0) and your localhost should be encrypted to protect any sensitive information.
+- It tests a very critical part of the development stack if you run your server on secure channels in production.
+- Cookies using secure or [same-site](https://auth0.com/blog/browser-behavior-changes-what-developers-need-to-know/) will not be sent to local `http`, disrupting authentication and other functionality.
```
```suggestion
- Cookies using a Secure attribute or [SameSite](https://auth0.com/blog/browser-behavior-changes-what-developers-need-to-know/) attribute set to `None` will not be sent across insecure channels, disrupting authentication and other functionality.
```

joshcanhelp commented on articles/libraries/secure-local-development.md:
```diff
@@ -0,0 +1,75 @@
+3. Serve the SSL certificate from your web application
+
+### 1. Install Mkcert Utility
+To get started, download [Mkcert](https://github.com/FiloSottile/mkcert) and follow the [installation instructions](https://github.com/FiloSottile/mkcert#installation) for your specific operating system.
```
```suggestion
To get started, download [mkcert](https://github.com/FiloSottile/mkcert) and follow the [installation instructions](https://github.com/FiloSottile/mkcert#installation) for your specific operating system.
```

joshcanhelp reviewed (changes requested):
Looks great. Couple of minor changes and I think we're GTG here.

davidpatrick committed:
Update articles/libraries/secure-local-development.md

Co-Authored-By: Josh Cunningham <josh.cunningham@auth0.com>

davidpatrick committed:
Update articles/libraries/secure-local-development.md

Co-Authored-By: Josh Cunningham <josh.cunningham@auth0.com>

davidpatrick committed:
Update articles/libraries/secure-local-development.md

Co-Authored-By: Josh Cunningham <josh.cunningham@auth0.com>

davidpatrick committed:
Update articles/libraries/secure-local-development.md

Co-Authored-By: Josh Cunningham <josh.cunningham@auth0.com>

davidpatrick committed:
Update articles/libraries/secure-local-development.md

Co-Authored-By: Josh Cunningham <josh.cunningham@auth0.com>
